src/api/code-challenge/content-types/code-challenge/utils.ts

```diff
@@ -0,0 +1,145 @@
+// Import filepaths changed from original
+import { removeJSComments } from "../code-challenge/curriculum-helpers";
+import { getErrorMessage } from "../code-challenge/general";
+
+type consoleLogOverride = (data: unknown[]) => void;
+
+export const overrideConsoleLog = (override: consoleLogOverride) => {
+  // @ts-expect-error will fix later
+  console.standardLog = console.log;
+  console.log = (...data: unknown[]) => {
+    override(data);
+    // @ts-expect-error will fix later
+    console.standardLog.apply(console, data); // eslint-disable-line prefer-spread
+  };
+};
+
+export const restoreConsoleLog = () => {
+  // @ts-expect-error will fix later
+  if (!console.standardLog) {
+    throw new Error(
+      "Attempted to restore console.log but it has never been overwritten"
+    );
+  }
+  // @ts-expect-error will fix later
+  console.log = console.standardLog;
+};
+
+interface IFindFirstPassingLineForCondition {
+  condition: string;
+}
+
+const getCodeEvaluationHelpers = (
+  logs: Array<Array<unknown>>,
+  codeString: string
+) => {
+  const helpers = {
+    // right now this only works with primitives
+    wasLoggedToConsole: (val: unknown) => {
+      return logs.some((logGroup) => logGroup.some((log) => log === val));
+    },
+    wasFunctionInvoked: (functionName: string) => {
+      return codeString.trim().includes(`${functionName}()`);
+    },
+    /**
+     * Evaluates a condition for each line of the provided code.
+     * Upon finding a line that passes it returns that line's index
+     */
+    findFirstPassingLineForCondition: ({
+      condition,
+    }: IFindFirstPassingLineForCondition) => {
+      debugger;
+      const splitCodeString = codeString.split("\n");
+      let codeToEvaluate = "";
+
+      for (let i = 0; i < splitCodeString.length; i++) {
+        const currentLine = splitCodeString[i];
+        codeToEvaluate += i === 0 ? currentLine : `\n${currentLine}`;
+        const context = getEvaluationContext(codeToEvaluate);
+        const codeWithCondition = `${codeToEvaluate}\n${condition}`;
+
+        if (evaluateWithContext(codeWithCondition, context)) {
+          return {
+            index: i,
+            passed: true,
+          };
+        }
+      }
```

Comment:
I honeslty think it'd be a bit more intuitive to read if you put `evaluateWithContext(codeWithCondition, context)` in a variable

Reply:
This is yours, but I tried to apply your suggestion.

Reply:
jesus fuck

Reply:
yeah I'm not setting the best example sorry.

Reply:
Although as far as examples go, my rushed code is still decent. Just not as good as my normal stuff